cluster-controller/src/main/java/io/strimzi/controller/cluster/operations/resource/AbstractOperations.java

```diff
@@ -84,74 +104,51 @@ public AbstractOperations(Vertx vertx, C client, String resourceKind) {
-        Future<Void> fut = Future.future();
-        vertx.createSharedWorkerExecutor("kubernetes-ops-pool").executeBlocking(
-            future -> {
-                if (operation().inNamespace(namespace).withName(name).get() != null) {
-                    try {
-                        log.info("Deleting {} {} in namespace {}", resourceKind, name, namespace);
-                        operation().inNamespace(namespace).withName(name).delete();
-                        log.info("{} {} in namespace {} has been deleted", resourceKind, name, namespace);
-                        future.complete();
-                    } catch (Exception e) {
-                        log.error("Caught exception while deleting {} {} in namespace {}", resourceKind, name, namespace, e);
-                        future.fail(e);
-                    }
-                } else {
-                    log.warn("{} {} in namespace {} doesn't exist, so cannot be deleted", resourceKind, name, namespace);
-                    future.complete();
-                }
-            }, false,
-            fut.completer()
-        );
-        return fut;
+    protected Future<ReconcileResult<P>> internalDelete(String namespace, String name) {
+        try {
+            log.info("Deleting {} {} in namespace {}", resourceKind, name, namespace);
+            operation().inNamespace(namespace).withName(name).delete();
+            log.info("{} {} in namespace {} has been deleted", resourceKind, name, namespace);
+            return Future.succeededFuture(ReconcileResult.deleted());
+        } catch (Exception e) {
+            log.error("Caught exception while deleting {} {} in namespace {}", resourceKind, name, namespace, e);
+            return Future.failedFuture(e);
+        }
     }
 
     /**
-     * Asynchronously patch the resource with the given {@code name} in the given {@code namespace}
-     * to reflect the state given in the {@code patch}, returning a future for the outcome.
-     * The patch cascades to related resources.
-     * @param namespace The namespace of the resource to patch.
-     * @param name The name of the resource to patch.
-     * @param patch The desired state of the resource.
+     * Patches the resource with the given namespace and name to match the given desired resource
+     * and completes the given future accordingly.
      */
-    public Future<Void> patch(String namespace, String name, T patch) {
-        return patch(namespace, name, true, patch);
+    protected Future<ReconcileResult<P>> internalPatch(String namespace, String name, T current, T desired) {
+        try {
+            log.info("Patching {} resource {} in namespace {} with {}", resourceKind, name, namespace, desired);
+            operation().inNamespace(namespace).withName(name).cascading(true).patch(desired);
+            log.info("{} {} in namespace {} has been patched", resourceKind, name, namespace);
+            return Future.succeededFuture(ReconcileResult.patched(null));
```

Comment:
Yes, but it's ugly: It would be possible to reproduce the logic that Fabric8 uses to construct the patch (serialize both resources to JSON, construct a JSON patch) and see if it's empty (or if it contains particular JSON paths). I considered that instead of the approach used in `Diff`, but rejected it because I felt it was horrid to do that only for Fabric8 to go and do the same thing.